src/Symfony/Component/DependencyInjection/Secrets/JweHandler.php

```diff
@@ -0,0 +1,291 @@
+            }
+
+            if (file_exists($this->privateKeyLocation)) {
+                $requiredOverwrites[] = $this->privateKeyLocation;
+            }
+
+            if (!empty($requiredOverwrites)) {
+                throw new SecretsOverwriteRequiredException(sprintf(
+                    'secrets files at %s already exist',
+                    implode(', ', $requiredOverwrites)
+                ), $requiredOverwrites);
+            }
+        }
+
+        $this->writeFile(
+            $this->secretsLocation,
+            json_encode(array(), self::JSON_ENCODE_OPTIONS),
+            $makeConfigDir = true
+        );
+
+        $this->writeNewKeyPair();
+    }
+
+    /**
+     * @throws InvalidArgumentException if the plaintext file does not contain valid secrets
+     */
+    public function regenerateEncryptedEntries(string $plaintextLocation): void
+    {
+        $this->validateConfig();
+        if ($this->isValidSecretsFile($plaintextLocation) &&
+            false !== $plaintextSecrets = $this->readSecrets($plaintextLocation)) {
+            $this->secrets = array();
+            foreach ($plaintextSecrets as $key => $secret) {
+                if (!preg_match('/^(?:\w++:)*+\w++$/', $key)) {
+                    throw new \InvalidArgumentException(sprintf(
+                        'The name for %s is invalid, secrets cannot be updated. Only "word" characters can be used in variable names.',
+                        $key
+                    ));
+                }
+
+                $this->secrets[$key] = JweEntry::encrypt($secret, $this->getPublicKey());
+            }
+        } else {
+            throw new InvalidArgumentException(sprintf(
+                'plaintext key location %s must point to a file with valid json secrets',
+                $plaintextLocation
+            ));
+        }
+
+        $this->writeEncrypted();
+    }
+
+    /**
+     * generates a new key pair, and writes it to the configured public and private key locations after re-encrypting secrets.
+     *
+     * @throws InvalidArgumentException if the $publicKeyLocation or $privateKeyLocation cannot be written
+     */
+    public function updateKeyPair(): void
+    {
+        $tempSecretsLocation = tempnam(sys_get_temp_dir(), '');
+        try {
+            //TODO should probably add a temp backup and make this behave like a transaction
+            $this->writePlaintext($tempSecretsLocation);
+            $this->writeNewKeyPair();
+            $this->regenerateEncryptedEntries($tempSecretsLocation);
+        } finally {
+            unlink($tempSecretsLocation);
+        }
+    }
+
+    /**
+     * @throws InvalidArgumentException if the secretsLocation does not point to a readable secrets file with valid json,
+     *                                  or if the public and private key files do not point to existing, readable files with 32 byte key values
+     */
+    public function validateConfig(bool $decryptRequired = false): void
```

Comment:
I still don't love this method. I'd rather it be split up, and included in other methods that a more directly influencing what we're trying to do at the time.

For example, most of the methods that call this ultimately want the `$this->secrets` property. So, instead of calling `$this->validateConfig()`, then `$this->populateSecrets()`, why not add a new `$this->getSecrets()` method? Inside, `if null === $this->secrets`, we would run all the checks needed to make sure that we can decrypt the secrets, and then return the secrets.

Here's another example: Move the check for a valid `$this->privateKeyLocation` into the `getPrivateKey()` method. Then, that validation will naturally occur only if & when the private key is actually needed.

Reply:
Yes that seems much cleaner, thank you. addressed in the latest commits